src/platform/w3dengine/client/w3dview.cpp

```diff
@@ -790,3 +797,962 @@ void W3DView::Move_Along_Waypoint_Path(int milliseconds)
+    float z2 = GameMath::Max(z1, g_theTerrainLogic->Get_Ground_Height(x + 40.0, y - 40.0, nullptr));
+    float z3 = GameMath::Max(z2, g_theTerrainLogic->Get_Ground_Height(x - 40.0, y - 40.0, nullptr));
+    float z4 = GameMath::Max(z3, g_theTerrainLogic->Get_Ground_Height(x + 40.0, y + 40.0, nullptr));
+    return GameMath::Max(z4, g_theTerrainLogic->Get_Ground_Height(x - 40.0, y + 40.0, nullptr));
+}
+
+void Draw_Drawable(Drawable *draw, void *user_data)
+{
+    draw->Draw(static_cast<View *>(user_data));
+}
+
+void W3DView::Update()
+{
+    static float follow_factor;
+    bool b1 = false;
+    bool b2 = false;
+
+    if (g_theTerrainRenderObject->Get_Need_Full_Update()) {
+        RefMultiListIterator<RenderObjClass> *iter = W3DDisplay::s_3DScene->Create_Lights_Iterator();
+        g_theTerrainRenderObject->Update_Center(m_3DCamera, iter);
+
+        if (iter != nullptr) {
+            W3DDisplay::s_3DScene->Destroy_Lights_Iterator(iter);
+            iter = nullptr;
+        }
+    }
+
+    ObjectID camera_lock = Get_Camera_Lock();
+
+    if (camera_lock == INVALID_OBJECT_ID) {
+        follow_factor = -1.0f;
+    }
+
+    if (camera_lock != INVALID_OBJECT_ID) {
+        m_doingMoveCameraOnWaypointPath = false;
+        m_doingMoveCameraAlongWaypointPath = false;
+        Object *obj = g_theGameLogic->Find_Object_By_ID(camera_lock);
+
+        if (obj != nullptr) {
+            if (follow_factor >= 0.0f) {
+                follow_factor += 0.05f;
+
+                if (follow_factor > 1.0f) {
+                    follow_factor = 1.0f;
+                }
+            } else {
+                follow_factor = 0.05f;
+            }
+
+            if (Get_Camera_Lock_Drawable() != nullptr) {
+                Drawable *drawable = Get_Camera_Lock_Drawable();
+
+                if (drawable != nullptr) {
+                    Matrix3D tm;
+                    Coord3D position;
+                    float radius;
+
+                    if (drawable->Client_Only_Get_First_Render_Obj_Info(&position, &radius, &tm)) {
+                        Vector3 v1(0.0f, 0.0f, 1.0f);
+                        Vector3 v2;
+                        v2.X = position.x;
+                        v2.Y = position.y;
+                        v2.Z = position.z;
+                        float f1 = radius * 1.0f;
+                        v2 += f1 * v1;
+                        Vector3 v3 = tm.Get_X_Vector();
+                        Vector3 v4 = v3 * radius;
+                        Vector3 v5 = v4 * 4.5f;
+                        Vector3 v6 = v2 - v5;
+                        Vector3 v7 = m_3DCamera->Get_Position();
+                        Vector3 v8 = v6 - v7;
+                        Vector3 v9 = v8 * 0.1f;
+                        v6 = v7 + v9;
+                        Matrix3D new_tm;
+                        new_tm.Look_At(v6, v2, 0.0f);
+                        m_3DCamera->Set_Transform(new_tm);
+                        b1 = false;
```

Comment:
This looks suspicious, because it has no effect.

Reply:
This is correct.